KafkaClusterRebalance CRD and rebalance operator

ppatierno opened the pull request:
### Type of change

- Enhancement / new feature

### Description

This PR adds the new `KafkaClusterRebalancer` custom resource and the related operator for handling it; it has it's own state machine to allow users to create a KCR asking for a rebalance proposal to Cruise Control and then actually executing it.
The KCR status is updated accordingly when the operator moves through the states machine.
We are still working on more ITs which will be available in a different PR.

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [ ] Update/write design documentation in `./design`
- [ ] Write tests
- [ ] Make sure all tests pass
- [ ] Update documentation
- [ ] Check RBAC rights for Kubernetes / OpenShift roles
- [ ] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [ ] Reference relevant issue(s) and close them after merging
- [ ] Update CHANGELOG.md

scholzj commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaClusterRebalanceAssemblyOperator.java:
```diff
@@ -0,0 +1,804 @@
+                    String clusterName = kafkaClusterRebalance.getMetadata().getLabels() == null ? null : kafkaClusterRebalance.getMetadata().getLabels().get(Labels.STRIMZI_CLUSTER_LABEL);
+                    String clusterNamespace = kafkaClusterRebalance.getMetadata().getNamespace();
+                    KafkaClusterRebalanceStatus desiredStatus = new KafkaClusterRebalanceStatus();
+                    if (clusterName != null) {
```
It looks like you execute this logic here only when event is triggered. Maybe I missed something, but this looks wrong. You have your own reconcileAll loop and you need to test this everytime in the loop, or?

ppatierno replied:
good catch!

scholzj commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaClusterRebalanceAssemblyOperator.java:
```diff
@@ -0,0 +1,804 @@
+                });
+    }
+
+    private RebalanceAnnotation getRebalanceAnnotation(KafkaClusterRebalance clusterRebalance) {
```
Some Javadocs would not hurt ... you have `getRebalanceAnnotation`, `rebalanceAnnotation`, hasRebalanceAnnotation` ... not exactly clear what the difference is between those.

ppatierno commented on api/src/main/java/io/strimzi/api/kafka/model/KafkaClusterRebalance.java:
```diff
@@ -0,0 +1,152 @@
+    public static final String SCOPE = "Namespaced";
+    public static final String V1ALPHA1 = "v1alpha1";
+    public static final List<String> VERSIONS = unmodifiableList(asList(V1ALPHA1));
+    public static final String RESOURCE_KIND = "KafkaClusterRebalance";
```
I thought about it for a long time, leaving it as it is but yes I would agree.

scholzj replied:
There is no going back ... so I think you should either agree and change it or disagree and leave it as it is :-o

ppatierno replied:
What I meant is that I agree, just asking to myself why I didn't do that before opening the PR. So, I will change!

scholzj replied:
Ahh, ok ... understood ... thanks.

ppatierno replied:
Done

ppatierno commented on api/src/main/java/io/strimzi/api/kafka/model/KafkaClusterRebalanceSpec.java:
```diff
@@ -0,0 +1,77 @@
+    }
+
+    @Override
+    public String toString() {
```
I see it in KafkaBridge, KafkaUser, KafkaMirrorMaker ... just for naming a few of them.

scholzj replied:
But matching classes for this class are KafkaUserSpec, KafkaBridgeSpec and KafkaMirrorMakerSpec. And I do not see it in any of those.

ppatierno replied:
Right

ppatierno commented on api/src/main/java/io/strimzi/api/kafka/model/status/KafkaClusterRebalanceStatus.java:
```diff
@@ -0,0 +1,55 @@
+        this.optimizationResult = optimizationResult;
+    }
+
+    @Description("The session identifier for requests to Cruise Control pertaining to this KafkaClusterRebalance resource.")
```
No, we are using the session in the status as a way to store this information that the operator needs for tracking the state of the rebalance.

scholzj replied:
Should that be somehow clarified in the description?

tomncooper replied:
I added a clarification about it use.

tomncooper commented on cluster-operator/pom.xml:
```diff
@@ -270,6 +298,12 @@
                 </includes>
             </resource>
+            <resource>
+                <directory>src/test/resources/io/strimzi/operator/cluster/operator/assembly/CruiseControlJSON</directory>
```
This is a folder full of example JSON files that are returned by the mock Cruise Control REST API. It is in the pom so that the MockServer instance in the test package can easily load them.

tomncooper replied:
Can we use `<scope>testing</scope>` with resource?

tomncooper replied:
So apprently things in the `src/test/resources` are ignored by the assembly plugin unless explictly included. If I add the fully qualilifed path to the resources in the MockServer they can be loaded without the need to have a specific resource plugin entry.

tombentley commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaClusterRebalanceAssemblyOperator.java:
```diff
@@ -0,0 +1,804 @@
+        if (rebalanceAnnotation == RebalanceAnnotation.none) {
+            log.info("{}: Arming Cruise Control rebalance user task status timer", reconciliation);
+            String sessionId = clusterRebalance.getStatus().getSessionId();
+            AtomicInteger ccAPIErrorCount = new AtomicInteger();
```
`ccApiErrorCount` would be more consistent with most of the rest of the project, I think. (Or have I lost this battle?)

tombentley commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaClusterRebalanceAssemblyOperator.java:
```diff
@@ -0,0 +1,804 @@
+        return rebalanceAnnotation;
+    }
+
+    private String rebalanceAnnotation(KafkaClusterRebalance clusterRebalance) {
+        return clusterRebalance.getMetadata().getAnnotations() == null ?
+                null : clusterRebalance.getMetadata().getAnnotations().get(ANNO_STRIMZI_IO_REBALANCE);
+    }
```
Do we really need this one?

ppatierno replied:
This takes the raw string value for the annotation which could even be a not valid value (i.e. `strimzi.io/rebalance=foo`). The other method maps to the corresponding annotation to `unknown` in this case (or `none` if no annotation is present). I renamed and added a few Javadoc on all of them.

tombentley commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/cruisecontrol/CruiseControlApi.java:
```diff
@@ -0,0 +1,25 @@
+    String CC_REST_API_USER_ID_HEADER = "User-Task-ID";
+    String CC_REST_API_SUMMARY = "summary";
+
+    Future<CruiseControlResponse> getCruiseControlState(String host, int port, boolean verbose);
+    Future<CruiseControlRebalanceResponse> rebalance(String host, int port, RebalanceOptions options, String userTaskID);
+    Future<CruiseControlResponse> getUserTaskStatus(String host, int port, String userTaskId);
+    Future<CruiseControlResponse> stopExecution(String host, int port);
```
Please can we document all of these?

tombentley commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/cruisecontrol/CruiseControlApiImpl.java:
```diff
@@ -0,0 +1,239 @@
+                    } else {
+                        result.fail(new CruiseControlRestException(
+                                "Unexpected status code " + response.statusCode() + " for POST request to " +
+                                host + ":" + port + path));
```
I think you should include the `statusMessage`. I also see scope here for some common error handling with the Kafka Connect client, but not for this PR. This comment applies to other similar places in this class.

tombentley commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/cruisecontrol/CruiseControlEndpoints.java:
```diff
@@ -0,0 +1,21 @@
+ */
+package io.strimzi.operator.cluster.operator.assembly.cruisecontrol;
+
+enum CruiseControlEndpoints {
```
Just a point to note about this pattern: an `enum` like this works fine when all your paths are static (as here), but not when the path contains some parameters, e.g. `/foo/${name}`.

tombentley commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/cruisecontrol/PathBuilder.java:
```diff
@@ -0,0 +1,68 @@
+        } else {
+            firstParam = false;
+        }
+        constructedPath += parameter;
```
Shouldn't you URL encode the value?

tombentley commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/cruisecontrol/PathBuilder.java:
```diff
@@ -0,0 +1,68 @@
+        } else {
+            firstParam = false;
+        }
+        constructedPath += param.asPair(value);
```
Shouldn't you URL encode the value?

tombentley commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/cruisecontrol/PathBuilder.java:
```diff
@@ -0,0 +1,68 @@
+        } else {
+            firstParam = false;
+        }
+        constructedPath += param.asList(values);
```
Shouldn't you URL encode the values?

tombentley commented on cluster-operator/src/test/resources/io/strimzi/operator/cluster/operator/assembly/CruiseControlJSON/CC-Rebalance-NotEnoughValidWindows-error.json:
```diff
@@ -0,0 +1 @@
+{"errorMessage":"Error processing POST request \u0027/rebalance\u0027 due to: \u0027com.linkedin.kafka.cruisecontrol.exception.KafkaCruiseControlException: com.linkedin.cruisecontrol.exception.NotEnoughValidWindowsException: There are only 0 valid windows when aggregating in range [-1, 1587403196210] for aggregation options (minValidEntityRatio\u003d1.00, minValidEntityGroupRatio\u003d0.00, minVa ...
```
Is the `\u0027` a java thing, or a json thing? Can't we just let the file be UTF-8 encoded and use the real value, rather than needing an escaped one?

tomncooper replied:
These files are the result of `curl <endpoint with params> > file.json` so are raw output that the mock server will return verbatim.

tombentley replied:
OK, that's fine then.

tombentley commented on operator-common/src/main/java/io/strimzi/operator/common/operator/resource/StatusUtils.java:
```diff
@@ -29,18 +29,22 @@ public static String iso8601Now() {
-                    .withType("Ready")
-                    .withStatus("True")
+                    .withType(type)
+                    .withStatus("Ready")
```
Did we mean to change the status here?

ppatierno replied:
You are absolutely right. It should be just "True" as it is today.

samuel-hawker commented on cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/cruisecontrol/CruiseControlClientTest.java:
```diff
@@ -0,0 +1,151 @@
+        client.getUserTaskStatus(HOST, PORT, userTaskID).setHandler(context.succeeding(result -> {
+            context.verify(() -> assertThat(result.getUserTaskId(), is(MockCruiseControl.USER_TASK_REBALANCE_NO_GOALS_VERBOSE_RESPONSE_UTID)));
+            context.verify(() -> assertThat(result.getJson().getJsonObject(CC_REST_API_SUMMARY), is(notNullValue())));
+            context.completeNow();
```
I generally prefer calling context.checkpoint() outside of the sync call and calling flag() inside this handler

samuel-hawker commented on cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/cruisecontrol/MockCruiseControlTest.java:
```diff
@@ -0,0 +1,169 @@
+            );
+            return Future.succeededFuture(response);
+        }).setHandler(context.succeeding(result -> {
+            context.completeNow();
```
Similarly can we use checkpoints here

samuel-hawker commented on cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/cruisecontrol/MockCruiseControlTest.java:
```diff
@@ -0,0 +1,169 @@
+            );
+            return Future.succeededFuture(response);
+        }).setHandler(context.succeeding(result -> {
+            context.completeNow();
```
ditto

samuel-hawker commented on cluster-operator/src/test/resources/io/strimzi/operator/cluster/operator/assembly/CruiseControlJSON/CC-State-proposal-not-ready.json:
```diff
@@ -0,0 +1 @@
+{"AnalyzerState":{"isProposalReady":false,"readyGoals":[]},"MonitorState":{"trainingPct":9.6,"trained":false,"numFlawedPartitions":0,"state":"RUNNING","numTotalPartitions":2065,"numMonitoredWindows":1,"monitoringCoveragePct":100.0,"reasonOfLatestPauseOrResume":"N/A","numValidPartitions":2065},"ExecutorState":{"state":"NO_TASK_IN_PROGRESS"},"AnomalyDetectorState":{"recentBrokerFailures":[],"recent ...
```
Is there a reason we haven't inlined these JSON strings into a java class?

tomncooper replied:
These files are example return response for the Mock Cruise Control sever and are copies of the raw response from `curl <cc-endpoint> > file.json` and are returned verbatim by the mock server.

ppatierno commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaClusterRebalanceAssemblyOperator.java:
```diff
@@ -0,0 +1,804 @@
+                @Override
+                public void eventReceived(Action action, KafkaClusterRebalance kafkaClusterRebalance) {
+                    log.debug("EventReceived {} on {} with status [{}] and {}={}", action, kafkaClusterRebalance.getMetadata().getName(),
+                            kafkaClusterRebalance.getStatus() != null ? kafkaClusterRebalance.getStatus().getConditions().get(0).getType() : null,
```
What do you mean? We want to print in which status is the rebalance operator state machine for printing a log.

scholzj replied:
Well, right now you are printing which ever condition is in the position 0. But you probably want to print status of some specific condition?

ppatierno replied:
Our conditions collection is always made by just one condition. Isn't it the case in every status field for all the resources we are handling today?

scholzj replied:
Some resources have more conditions - the Kafka one for example ... at least in some cases. That is why this looks a bit fragile - you have one condition now ... until you add more in some future and forget to change this.

ppatierno replied:
Anyway, there is not a type I should search for as you mentioned in the first comment. We need the only one condition we have and the corresponding type in order to log it.

ppatierno commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaClusterRebalanceAssemblyOperator.java:
```diff
@@ -0,0 +1,804 @@
+        return Future.succeededFuture(clusterRebalance);
+    }
+
+    enum State {
```
but arent't the api supposed to be used somehow by users as well? we provide them even on Maven right? so a user can develop its own application interacting with Strimzi using this API. In this case we are talking about internal stuff related to the rebalance operator state machine.

tombentley replied:
Yes, but the nature of the contract is different. If, in the `api` and also in the CRD JSON Schema it's an enum then users are entitled to conclude that items won't be added to the enumeration (e.g. users of the `api` module can write a `switch` over the enum with cases for each value). If we later need to add a value we're breaking that code. If it's just a String then they shouldn't be making those kinds of assumptions.

ppatierno replied:
So I am confused, I was suggesting to leave this enum here. It seems you are on the same page @tombentley ?

tombentley replied:
I think it's OK here, yes.

scholzj replied:
So if this state is internal only, it is not used in the custom resources then? My understanding was that this is in the Status of the CR and I expected that users are interested in this information -> E.g. the proposal is pending, I should wait more, the proposal is ready, I should review it. That seems like exactly the thing which the user wants to do using the `api`. If we pretend that the status in the CR is just opaque String, it will force the user to write their own Enum and they will have to for every release dig the possible states out of the docs and they will have no build level checks that the old states still exist. So if this is part of the status, it should be AFAIC part of the `api`.

tombentley replied:
The states should be public and documented, but we shouldn't imply that the ones which exist today will be the only ones. It's the exhaustivity which is the problem, not the visibility.

scholzj replied:
I'm not sure I really understand it ... it is public APi, users need to understand them and use them, but we make their life miserable by forcing the to dig it out of the docs and double check it every release?

On a practical level, the list of the states in docs is as exhaustive and final as Enum would be. It is just much harder to use.

ppatierno commented on api/src/main/java/io/strimzi/api/kafka/model/status/KafkaClusterRebalanceStatus.java:
```diff
@@ -0,0 +1,55 @@
+    @Description("A JSON describing the optimization result")
+    @JsonInclude(JsonInclude.Include.NON_EMPTY)
+    public Map<String, Object> getOptimizationResult() {
+        return optimizationResult;
```
There is no JSON that is passing back from the operator to the API, if I understood correctly your concern.

tombentley replied:
So what are the `Objects` in this `Map<String, Object>`?

tomncooper replied:
The `optimisationResults` are a `JsonObject` of String keys to String, double, int and lists of Strings. Hence the use of `Object`.

tombentley replied:
So you're saying that the lists of Strings are always going to be in the same order?

tomncooper replied:
No, but the order does not matter. The map is just given as is to the Status to shown to the user.

scholzj replied:
That turned out to be a bit more complicated in Mirror Maker 2. I wonder if it would end up the same here :-o.

kyguy commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/model/CruiseControl.java:
```diff
@@ -227,16 +237,64 @@ public static CruiseControl fromCrd(Kafka kafkaAssembly, KafkaVersion.Lookup ver
+        Set<String> anomalyDetectionGoals = new HashSet<>(Arrays.asList(anomalyGoalsString.split("\\s*,\\s*")));
+
+        String defaultGoalsString = configuration.getConfigOption(CRUISE_CONTROL_DEFAULT_GOALS_CONFIG_KEY);
+        Set<String> defaultGoals = new HashSet<>(Arrays.asList(defaultGoalsString.split("\\s*,\\s*")));
```
We set it to this [1] here [2] if default.goals is set to null


[1] https://github.com/strimzi/strimzi-kafka-operator/blob/master/cluster-operator/src/main/java/io/strimzi/operator/cluster/model/CruiseControlConfiguration.java#L26-L41
[2] https://github.com/strimzi/strimzi-kafka-operator/blob/e1cad47d1a4f70a812e14895e8d1eb09a3964095/cluster-operator/src/main/java/io/strimzi/operator/cluster/model/CruiseControl.java#L242-L244

tombentley replied:
Which is great and true _today_, but there's no harm in adding a null check in case it's not true _tomorrow_. Remember, the more distant the place where you know the value is always set then the more fragile it is to a future change violating the non-nullness.

kyguy commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/model/CruiseControl.java:
```diff
@@ -227,16 +237,64 @@ public static CruiseControl fromCrd(Kafka kafkaAssembly, KafkaVersion.Lookup ver
+            }
+        }
+
+        // Either self.healing.goals is null or it is an empty list
```
This is explaining the result of the code above, we could probably move it up there

tomncooper commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/model/CruiseControlConfiguration.java:
```diff
@@ -23,7 +23,7 @@
      * The high priority goals will be executed first.
      */
-    private static final String DEFAULT_GOALS = "com.linkedin.kafka.cruisecontrol.analyzer.goals.RackAwareGoal," +
+    private static final String GOALS = "com.linkedin.kafka.cruisecontrol.analyzer.goals.RackAwareGoal," +
```
In the deployment PR the default `goals` list was built from the default `default.goals` list. This was the wrong way round so this PR swaps them.

kyguy replied:
`default.goals` must be a subset of `goals` so we thought it would be more intuitive sense

tomncooper commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaClusterRebalanceAssemblyOperator.java:
```diff
@@ -0,0 +1,804 @@
+                                log.info("{}: Getting Cruise Control rebalance user task status", reconciliation);
+                                // TODO: If this request results in multiple errors from the server then the Rebalance state
+                                //       will switch to NotReady -> ProposalPending -> ProposalReady even though we are in
+                                //       an ongoing rebalance. We may need to add an Error state to the state machine.
```
This has actually been fixed now by the `onNotReady` method. So you can remove this @ppatierno

ppatierno replied:
Yep.

ppatierno commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaClusterRebalanceAssemblyOperator.java:
```diff
@@ -0,0 +1,804 @@
+        Ready
+    }
+
+    enum RebalanceAnnotation {
```
I guess that the @tombentley 's comment on `State` is valid here as well. I would leave it here.

scholzj replied:
Same as above. Isn't this the value to which the user should set the annotation to tell CC what to do?

ppatierno replied:
Well, it's something like the "strimzi.io/delete-pod-and-pvc" and it's not in the `api` package, or?

scholzj replied:
Isn't the only valid value in `strimzi.io/delete-pod-and-pvc` always `true`? 

I leave it up to you guys, not a fight worth fighting. But I don't understand the difference this makes from the API perspective? If adding a new value is breaking, we should avoid regardless of Enum or docs. And if someone actually uses the API the Enum would be useful.

ppatierno replied:
it's always true but it's not in the api, it's just documented. To be honest from my point of view what is in the API is something usable from a user from code, not something used to interact with the custom resources.

scholzj replied:
The hole API is about custom resources and working with them from Java. So reading the status, processing it, triggering actions etc. is IMHO part of it.

ppatierno replied:
If we are saying that some user could write a Java application and interacting with the rebalance operator from there, then yes now I can see your point.
Anyway again, even if the delete pod and pvc is true, it should be there helping the user to know that it exists and not putting a raw string in his Java code.

scholzj replied:
For me that is something what makes sense ... for example writting an applicaiton which would create the KafkaRebalance resource with some goals, wait until the proposal is ready, check the proposal and trigger its application in the cluster. TBH, not sure how many users does the `api` have. Maybe its wasted effort.

tomncooper commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaClusterRebalanceAssemblyOperator.java:
```diff
@@ -0,0 +1,804 @@
+                                                            .withOptimizationResult(taskStatusJson.getJsonObject("summary").getMap())
+                                                            .addNewCondition().withType(State.Rebalancing.toString()).endCondition().build());
+                                                }
+                                                ccAPIErrorCount.set(0);
```
I did have one originally then thought it might be redundant. That count is for unsuccessful attempts to contact that CC server. If we have got here then we have successfully communicated and count can be reset.

ppatierno commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaClusterRebalanceAssemblyOperator.java:
```diff
@@ -0,0 +1,804 @@
+        });
+    }
+
+    public static Future<KafkaClusterRebalance> updateStatus(KafkaClusterRebalance clusterRebalance,
```
No it doesn't need to be public, changed to private.

ppatierno commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/cruisecontrol/CruiseControlApi.java:
```diff
@@ -0,0 +1,25 @@
+ * Copyright Strimzi authors.
+ * License: Apache License 2.0 (see the file LICENSE or http://apache.org/licenses/LICENSE-2.0.html).
+ */
+package io.strimzi.operator.cluster.operator.assembly.cruisecontrol;
```
I agree but at this point, we should move the `KafkaConnectApi` class as well under the `resource`. Of course, it's not related to the current PR.

tomncooper commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaClusterRebalanceAssemblyOperator.java:
```diff
@@ -0,0 +1,804 @@
+                // Check that we have not already failed to contact the API beyond the allowed number of times.
+                if (ccAPIErrorCount.get() >= MAX_API_RETRIES) {
+                    vertx.cancelTimer(t);
+                    p.fail(new RuntimeException("Unable to reach Cruise Control API after " + MAX_API_RETRIES + " attempts"));
```
Agree that would be a good idea. This could be a good thing to do when we switch from the deprecated vertx httpclient [post](https://vertx.io/docs/apidocs/io/vertx/core/http/HttpClient.html#post-int-java.lang.String-java.lang.String-io.vertx.core.Handler-) and [get](https://vertx.io/docs/apidocs/io/vertx/core/http/HttpClient.html#get-java.lang.String-java.lang.String-io.vertx.core.Handler-) methods we are using and switch to the [webclient](https://vertx.io/docs/vertx-web-client/java/) for the Connect and CC clients.

ppatierno commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaClusterRebalanceAssemblyOperator.java:
```diff
@@ -0,0 +1,804 @@
+                        } else {
+                            log.debug("{}: Rebalance resource was deleted, stopping the request time", reconciliation);
+                            vertx.cancelTimer(t);
+                            p.complete();
```
In this specific case it's not called because the caller check that the resource is not there anymore so there is no status to update but you are right, I need to check null status in the `updateStatus` for another case, or in general. thanks.

ppatierno commented:
@tombentley @scholzj @samuel-hawker I think we addressed all your comments. The open points should be just questions that I, @kyguy and @tomncooper asnwered.
Maybe you could have another look at the PR. I will try to figure out what is wrong with Travis meanwhile. thanks.

scholzj commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaRebalanceAssemblyOperator.java:
```diff
@@ -0,0 +1,869 @@
+
+    private Future<Void> reconcileRebalance(Reconciliation reconciliation, KafkaRebalance kafkaRebalance) {
+        if (kafkaRebalance == null) {
+            log.info("{}: Rebalance resource deleted", reconciliation);
```
So ... shouldn't thi have any impact on things? Such as stopping the rebalance or proposal preparation?

ppatierno replied:
This code is mostly related to the trigger periodic reconcile. When a resource is deleted in the middle of a proposal computation or actual rebalancing, the resource deletion is handled there stopping polling mechanism and ongoing task. See `onPendingProposal` and `onRebalancing`.

scholzj commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaRebalanceAssemblyOperator.java:
```diff
@@ -0,0 +1,869 @@
+
+        return computeNextStatus(reconciliation, host, apiClient, clusterRebalance, currentState, rebalanceAnnotation, rebalanceOptionsBuilder)
+                .compose(desiredStatus -> {
+                    // due to a long rebalancing operation that takes the lock for the entire period, more events related to resource modification could be
```
Wait, this sounds confusing ... this for sure does nto work synchronously but is checking the states with Cruise Control asynchronously at every reconciliation, or? So what entire period does the lock need to be hold?

ppatierno replied:
When a rebalance resource is created, during the pending proposal computation and rebalancing a lock is hold on the resource with a timer/polling mechanism to check the action status on the CC REST API. It's released when the proposal is ready or the rebalancing is end. I did a lot of back and forth with @tombentley 3 months ago about doing this across reconciliation or in one reconciliation holding the lock. I should get our discussion but this came as the best solution. Even in this case, changing could mean starting from scratch.

scholzj replied:
I don't really understand this. You have periodic reconciliation -> when it is started you check the CC API and update the resource. Why is there any need for an additional timer?

scholzj replied:
What is the impact on resource consumption of this? We have already problem with this and I can immagine when many more addiitonal recocnile loops are added by this it will further increase?

ppatierno replied:
the first bad thing I remember is that you should wait for 2 minutes to have the resource updated but it's just a minor thing. The discussion was long over Christmas, my memory is bad and I should get the discussion content if possible with @tombentley :-)

scholzj replied:
Not sure how the 2 minutes are a problem when the rebalance might AFAIU take hours or days. Anyway, probably too late to change now.

tombentley replied:
* How long the proposal takes will depend on many factors. I don't think we really have enough experience of the parameter space to have any intuition, but I think days is unlikely. Seconds, minutes and hours all seem likely.
* But using a 2nd timer we are, in effect, optimising the <2 minutes case. We can form a judgement about whether this is the right way to do it when we have more experience.
* Whether a 2nd timer takes up much resources is something else where we have no data. 

So I think it's better to keep it as it is and be open to revising this later when we have more of a feeling for these things. Another option would be to add some websocket notification endpoint to CC so it could tell us without our needing to poll.

scholzj replied:
You really thing that majority of rebalances take less than 2 minutes? If I follow the idea that we do not have the data, then the simplest solution should have been chosen first and that is using the regular reocnciliation loop. I think that if nothing else this would cause lot of confusion around the locks.

ppatierno replied:
The "simplest" solution (as you call it) was tried and it didn't work well. Tbh it was 3 months ago and I don't remember the reason. Maybe it was my fault maybe not. Right now I think that this works and we can live with it, exploring and trying to use the reconcile loop with the next iteration in a future release (after 0.18.0).

scholzj commented on examples/kafka-rebalance/kafka-rebalance.yaml:
```diff
@@ -0,0 +1,11 @@
+apiVersion: kafka.strimzi.io/v1alpha1
```
Is the `kafka-rebalance` really what the users will be looking for?I'm not sure it says enougn. Why not just `cruide-control`. If nothing else I would cut the `kafka-` part since almost everything there is `kafka-`. But I think `cruide-control` or for example `cluster-rebalancing` would be more eye catching. A user who donwloads the examples and will see it will immediately know that we have something in this area.

ppatierno replied:
I was following the same pattern as for example user and topic ... we have `kafka-topic.yaml` and `kafka-user.yaml` example files, so `kafka-rebalance.yaml` seems to be fine to me.

scholzj replied:
The comment was more about the directory than the file it self. I think the file name is fine.

ppatierno replied:
Yeah I agree, changing the folder name is a good idea.

ppatierno commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaRebalanceAssemblyOperator.java:
```diff
@@ -0,0 +1,869 @@
+ * <p>
+ *     This operator takes care of the {@code KafkaRebalance} custom resources that a user can create in order
+ *     to interact with Cruise Control REST API and execute a cluster rebalancing.
+ *     A state machine is used for the rebalacing flow which is reflected in the {@code status} of the custom resource.
```
The `type` field in the status condition reflects the corresponding state of the rebalacing resource/state machine operator. Should I clarify it here?

scholzj replied:
But that means there is zero documentation for this in the API reference. I think you should clarify it here but also add it to the `api` - for example using `@DescriptionFile`

ppatierno replied:
I will add `@DescriptionFile` when doc will be available in `documentation/book/api` but in a next PR.

scholzj replied:
When would that be? That sounds like 0.19 only?

ppatierno replied:
@PaulRMellor @laidan6000 ?

scholzj replied:
Any changes to the `api` module should not be done without a further RC - so unlike the documentation or ST PRs which we historically merged without RCs, this would need an RX.

ppatierno commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaRebalanceAssemblyOperator.java:
```diff
@@ -0,0 +1,869 @@
+ *     to the {@code ProposalReady} state. The status of the {@code KafkaRebalance} custom resource is updated with the
+ *     computed rebalance proposal so that the user can view it and making a decision to execute it or not.
+ *     For starting the actual rebalancing on the cluster, the user annotate the custom resource with
+ *     the {@code strimzi.io/rebalance=approve} annotation, triggering the operator to send a rebalance request to the
```
We had a discussion in December about more or less how to handle this and the annotation was mentioned here https://github.com/strimzi/strimzi-kafka-operator/issues/2202 . At the end, it is like we interact to delete pod for example and annotation seems to be the more natural way instead of a field. To be honest I think that's too late for changing this now. Why do you think the field would be better than an annotation?

scholzj replied:
The main reason why we use the annotation for deleting a pod is that there we need it to disappear after the action. So using a field inside the `.spec` would mean that we have to delete it somehow. That does not necessarily apply here. But its fine if you thought about it and think that anno is better.

ppatierno replied:
Anyway, the operator deletes the annotation when the operation is executed.

ppatierno commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaRebalanceAssemblyOperator.java:
```diff
@@ -0,0 +1,869 @@
+ *     to the {@code Ready} state when the rebalancing is done.
+ *
+ *     The user is also able to stop an in-progress rebalance proposal computation or an actual rebalancing,
+ *     annotating the custom resource with {@code strimzi.io/rebalance=stop} when it is in {@code PendingProposal}
```
So stopping a in-progress rebalance proposal is not really a thing. There is no way on the CC REST API stopping something like that, stopping here is useful to stop the periodic time for requesting status to CC REST API. About rebalancing, yes there is a way for stopping an ongoing rebalancing task and I guess that it leaves the work as it is. Is that right @tomncooper ?

scholzj replied:
> So stopping a in-progress rebalance proposal is not really a thing.

This Javadoc says it does - it says you can use this annotation when the state is `Rebalancing`.

tomncooper replied:
So you can stop an on going `rebalance(dryryun=false)`, you cannot stop the creation of a optimization proposal (`rabalance(dryrun=true)`). As Paolo said the stop annotation here is to signal to the operator that you don't want it to hang around for the proposal to be ready.

scholzj replied:
I guess now I'm totally confused ... you say _cannot stop the creation of a optimization proposal_ but also that it is _to signal to the operator that you don't want it to hang around for the proposal to be ready_  ... isn't creation of the proposal and waiting for the proposal to be ready the same thing?

tomncooper replied:
So a user deploys a `rebalance.yaml` with a bunch of custom options. They apply it and off CC goes calculating the optimization while we hold the lock. The user then realises "dam I missed something" and stops the rebalance. If we did not allow for stopping a proposal (on the Strimzi side --- it is not possible on the CC side) the CO would wait for the custom proposal to be ready and then wait for the next reconcile before stopping things and releasing the lock.

scholzj replied:
Ok, that sounds reasonable. So perhaps you need to clarify here (+ somewhere on the docs, but that migth be other PR) what does do something in CC and what does just do something in the operator since that does not seem to be clear. Even now I still don't understand whether stopping an active rebalance is now again operator only or if that is actually in CC.

ppatierno replied:
stopping an active rebalance is an actual operation on CC

scholzj replied:
Ok ... since you are using single annotation for two different things, could you make it clear in the comment? (and in the docs?)

tomncooper replied:
I updated the comments with clarification of the two stop conditions.

ppatierno commented on api/src/main/java/io/strimzi/api/kafka/model/KafkaRebalanceSpec.java:
```diff
@@ -0,0 +1,65 @@
+
+    @Description("A list of goals, ordered by decreasing priority, to use for generating and executing the rebalance proposal." +
+            "The supported goals are available at https://github.com/linkedin/cruise-control#goals." +
+            "If an empty goals list is provided, the goals declared in the default.goals Cruise Control configuration parameter are used.")
```
> Are you 100% sure there will be never any use case for an actual empty list of goals?

Why are you asking this? I didn't say that, or I misunderstood what you mean :-)

scholzj replied:
It is a question if you are sure abotu it. The API implies so. You say that _empty list does actually not mean empty list but list of default goals_. I can live with it, but I'm saying that there is no going back from that. So you need to be sure that there will never be any reason why you would need the empty list to mean empty list.

tombentley replied:
Yeah, there is potentially a difference between not specifying the goals at all vs. specifying an empty list. If you're wanting to treat them as the same thing that's fine, but it's not a decision you can change later. From my understanding it wouldn't make sense to ask for a rebalance with no goals, so I think it's fine to conflate the two.

ppatierno replied:
Tbh no `goals` field and empty `goals` field sound to be the same to me. In both cases, I am not specifying goals. We should be just clear that "no goals" means using the default ones which is already described. Maybe just adding "If no goals or an empty goals list .... " . Is it what you mean?

scholzj replied:
I think there is a big difference between when something doesn't exist and when it exists and it is empty. As I said, if you are sure that you will never need another meaning for empty list it should be fine.

ppatierno replied:
In this case, specifying no `goal` and an empty `goals` list is semantically the same.

scholzj commented:
Given that I now understand better how this has several recocniliation loops - how does it deal with the operator metrics?

scholzj commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaRebalanceAssemblyOperator.java:
```diff
@@ -0,0 +1,869 @@
+        if (rebalanceAnnotation == RebalanceAnnotation.none) {
+            log.debug("{}: Arming Cruise Control rebalance proposal request timer", reconciliation);
+            vertx.setPeriodic(REBALANCE_POLLING_TIMER_MS, t -> {
+                resourceOperator.getAsync(clusterRebalance.getMetadata().getNamespace(), clusterRebalance.getMetadata().getName()).setHandler(getResult -> {
```
This is fairly heavy use of the Kubernetes API if you poll this every few seconds. I wonder if this will cause any issues.

tombentley replied:
It's a valid point. One thing we could do would be to utilise the watch we've already got to store the latest version when it changes and here to retrieve that. A little care other threading would be needed, but it wouldn't put any extra load on k8s.

ppatierno replied:
It could make sense but it means having a globally available `updatedRebalance` variable updated by the watcher every time it's fired inside a synchronized block and before acquiring the Vert.x lock for doing the reconcile. The above code could avoid calling K8S and using that global `updatedRebalance` in a sync block as well. My question is: if the watcher is raised more time, the global var will always contain the last updated one, but when the watcher code will be unlocked on the Vert.x lock for doing the reconcile, should it process the latest updated  or the one got when it was fired? I mean an old watch could operate on a more recent resource compared to when the watcher was raised.

ppatierno replied:
Anyway it needs more testing, I would leave it for later when we'll have more tests with upcoming PR.

ppatierno commented:
@scholzj so the operator metrics weren't there when this stuff was developed at the time. To be honest, I didn't think about it, my first thought is that metrics will miss the "internal reconcile", I should take a deep look at them to provide a better answer.

scholzj commented:
Maybe it is good if they miss it because that would cause complete chaos if you mix some internal 5 second recocniles with the regular operator reconciles. But you should look at it and make sure it works in some reasonable way.

I also wonder how many failed locks will your internal reconcile be triggering and confusing the  users.

tombentley commented on api/src/main/java/io/strimzi/api/kafka/model/KafkaRebalanceSpec.java:
```diff
@@ -0,0 +1,65 @@
+        this.goals = goals;
+    }
+
+    @Description("Whether to allow hard goals to be skipped in rebalance proposal generation. Default is false.")
```
Out of interest what's the use case for skipping hard goals. On the face of it skipping a hard goal is worse than skipping a soft goal. It might be worth mentioning this in the `@Description`.

ppatierno replied:
Of course, you should always hit hard goals but it's anyway a feature that CC allows to you, so why not exposing it?

tombentley replied:
But what's the use case for it? If we don't understand, how can we inform users about when to use this?

tomncooper replied:
Imagine by default you care about 4 particular hard goals and this is fine most of the time. However, you notice that your cluster has got into a really strange state (somehow) and that one goal will fix this specific issue. You want to fix only this issue and don't want the other configured `hard.goals` overriding it or preventing a solution being found. In that case you can `skip_hard_goal_check` and just get a proposal for the goal(s) you care about.

tombentley replied:
So why would I not just comment out those hard goals in my CR and do the rebalance? Why this special config for ignoring some other config?

ppatierno replied:
I could be wrong, @tomncooper and @kyguy can help on this, but maybe because there are some default hard goals that you don't have in your CR but they are there by default from CC default configuration and in the scenario Tom mentioned you are able to skip checks in this way.

tomncooper replied:
Default `hard.goals` are set at the CC server level. If you want to use your own set of goals in a `KafkaRebalance` CR they must include all those configured `hard.goals` or you will get an error. Unless you pass `skip_hard_goals=true`.

If you do not include any goals in a `KafkaRebalance` CR then CC will use the configured `default.goals` (also set at CC server level) to prepare the proposal.

tombentley replied:
OK, I understand now! Thanks! Can we turn that into a pithy sentence or two for the `@Description`? Something like

```java
@Description("Whether to allow the hard goals specified in the Kafka CR to be skipped in rebalance proposal generation. " + 
             "This can be useful when some of those hard goals are preventing a balance solution being found. " + 
             "Default is false.")
```

tombentley commented on api/src/main/java/io/strimzi/api/kafka/model/status/KafkaRebalanceStatus.java:
```diff
@@ -0,0 +1,57 @@
+    private String sessionId;
+    private Map<String, Object> optimizationResult = new HashMap<>(0);
+
+    @Description("A JSON describing the optimization result")
```
```suggestion
    @Description("A JSON object describing the optimization result")
```

tombentley commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/model/CruiseControl.java:
```diff
@@ -226,16 +236,56 @@ public static CruiseControl fromCrd(Kafka kafkaAssembly, KafkaVersion.Lookup ver
+        // Remove all the goals which are present in the default goals set from the anomaly detection goals
+        anomalyDetectionGoals.removeAll(defaultGoals);
+
+        if (anomalyDetectionGoals.size() > 0) {
```
```suggestion
        if (!anomalyDetectionGoals.isEmpty()) {
```

tombentley commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaRebalanceAssemblyOperator.java:
```diff
@@ -0,0 +1,869 @@
+ *     Cruise Control REST API in order to execute the rebalancing.
+ *     During the rebalancing, the operator state machine is in the {@code Rebalancing} state and it moves finally
+ *     to the {@code Ready} state when the rebalancing is done.
+ *
```
```suggestion
 * <pre><code>
 *   User        Kube           Operator              CC
 *    | Create KR  |               |                   |
 *    |----------->|   Watch       |                   |
 *    |            |-------------->|   Proposal        |
 *    |            |               |------------------>|
 *    |            |               |   Poll            |
 *    |            |               |------------------>|
 *    |            |               |   Poll            |
 *    |            | Update Status |------------------>|
 *    |            |<--------------|                   |
 *    |            |   Watch       |                   |
 *    |            |-------------->|                   |
 *    | Get        |               |                   |
 *    |----------->|               |                   |
 *    |            |               |                   |
 *    | Approve    |               |                   |
 *    |----------->|  Watch        |                   |
 *    |            |-------------->|   Rebalance       |
 *    |            |               |------------------>|
 *    |            |               |   Poll            |
 *    |            |               |------------------>|
 *    |            |               |   Poll            |
 *    |            | Update Status |------------------>|
 *    |            |<--------------|                   |
 *    |            |   Watch       |                   |
 *    |            |-------------->|                   |
 *    | Get        |               |                   |
 *    |----------->|               |                   |
 * </code></pre>
```

ppatierno replied:
Do you really want me to put this asciiart in the doc? Why not having official doc for this? Isn't it too much for a source file?

tombentley replied:
Sure, why not? Don't you think it summaries the basic flow in a readily understandable form? Much more easily understood than having to read paragraphs of text, imho. And I didn't know there could be _too much_ documentation in a place like this were only us developers are going to benefit from it.

ppatierno replied:
Ok, I added it.

tombentley replied:
You might thank me in 6 months time when you're having to debug a problem and can't remember how this code works :slightly_smiling_face:

tombentley commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaRebalanceAssemblyOperator.java:
```diff
@@ -0,0 +1,869 @@
+    private final PlatformFeaturesAvailability pfa;
+    private final Function<Vertx, CruiseControlApi> cruiseControlClientProvider;
+
+    private String ccHost = null;
```
```suggestion
    private final String ccHost = null;
```

tombentley commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaRebalanceAssemblyOperator.java:
```diff
@@ -0,0 +1,869 @@
+     * @param vertx The Vertx instance
+     * @param pfa Platform features availability properties
+     * @param supplier Supplies the operators for different resources
+     * @param ccHost Optional host address for the Cruise Control REST API. If this is not supplied then Cruise Control
+     *             service address will be used. This parameter is intended for use in testing.
```
Can you do this by:

* Having a protected method to return the `ccHost()` at the point you need it. Impl in this class returns the service address.
* Use a subclass for the test which returns something else? 

That would be neater than having this field which we don't actually need in the production code.

tombentley commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaRebalanceAssemblyOperator.java:
```diff
@@ -0,0 +1,869 @@
+                                                      Throwable e) {
+        if (e != null) {
+            StatusUtils.setStatusConditionAndObservedGeneration(clusterRebalance, desiredStatus, e);
+        } else if (desiredStatus.getConditions() != null && desiredStatus.getConditions().get(0).getType() != null) {
```
I still feel, like Jakub, that this is fragile. Let's do the right thing and fnd the condition that we're interested in rather than assuming that there is a nonempty list of conditions and that the one we want is the first.

ppatierno replied:
I still don't understand what you mean guys. Which condition should I search for? What should be the query and the parameter? There is only one condition, that's it. Can you elaborate please?

tombentley replied:
Today there's a single condition with `type: Status`. If the conditions are not empty we know it must be at index 0. We're not checking for emptiness, but if our code is perfect maybe the state machine guarantees it's never empty. But maybe our code is not perfect. If we later decided to add another condition with `type: Warning` then we can't be sure so certain it'll be at index 0 any more. So let's write a method which searches the conditions for the first condition of `type: Status`, with null checks on all the getters. It's just more robust code.

ppatierno replied:
Maybe this is the misunderstanding! `type` is not `Status`, but `type` contains the actual status ... so `type: PendingProposal`, `type: ProposalReady`, `type: Rebalancing` ... so we are storing the state machine status in the `type` field of the only one condition we have there.

tomncooper replied:
I think Tom means that in future there could be more than one condition and that we cannot guarantee that the CC state condition will be at position 0 in the list returned by `getConditions()`. If there was ever more than one condtion we could end up checking the type of a different condition.

We should have a function that loops through the condition list and searches for one that has a status type matching any of those defined in the state enum and then return that type.

ppatierno replied:
What's the value of that? At this point, we should completely change the structure of the status and set `type: State` then adding a new field `state` with the actual value as `PendingProposal`, `Ready`, .... while today something like `Ready` is just set in the `type` field.

tomncooper replied:
We don't meed to do that. This issue is what if `desiredStatus.getConditions().get(0)` is for a different condition where `getType()` is null but `desiredStatus.getConditions().get(1).getType()` is our condition and is not null?

We just need something like `findCCStateCondition(desiredStatus.getConditions()).getType() != null` or something similar.

samuel-hawker replied:
I'm in agreement here, this may well be fragile. implementing a method like Tom described should fix this.

tomncooper replied:
I have added a function to search for a condition which has a type matching one of the enum states and return that. This should prevent issues with multiple conditions in the future.

ppatierno replied:
We are using the `type` field as the state of the state machine, what's the filter we should use to get THE condition? I still don't understand. It makes sense to me if having more conditions, one has `type: State` so filtering on it and then getting the state from a `state` field but it's not the case here.

ppatierno replied:
We addressed this in the latest commit. Now `type` is always `State` in the condition, while the `status` field contains the state machine state (i.e. PendingProposal, Rebalancing, Ready and so on).

tombentley commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaRebalanceAssemblyOperator.java:
```diff
@@ -0,0 +1,869 @@
+        }
+        StatusDiff diff = new StatusDiff(clusterRebalance.getStatus(), desiredStatus);
+        if (!diff.isEmpty()) {
+            KafkaRebalance copy = new KafkaRebalanceBuilder(clusterRebalance).build();
+            copy.setStatus(desiredStatus);
```
Surely the idiomatic way to do this is to set the status in the builder, rather than build and then set? Then you can inline the whole builder expression and you only need one line where you currently have 3.

tombentley commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaRebalanceAssemblyOperator.java:
```diff
@@ -0,0 +1,869 @@
+                                            .compose(updatedClusterRebalance -> {
+                                                log.info("{}: State updated to [{}] with annotation {}={} ",
+                                                        reconciliation,
+                                                        updatedClusterRebalance.getStatus().getConditions().get(0).getType(),
```
Similar to the previous comment, this looks like an NPE waiting to happen. However you obtain the condition, you should factor this chain of `get`s into a method.

ppatierno replied:
The status was updated at this point with at least/max one condition, so how the NPE could happen? Maybe it's related to the fact that I haven't got yet what you mean in the previous comment. :-)

tombentley replied:
My point is not that the code is _broken_, but that it's _fragile_ to future changes.

tombentley commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaRebalanceAssemblyOperator.java:
```diff
@@ -0,0 +1,869 @@
+                log.debug("{}: No {} annotation set", reconciliation, ANNO_STRIMZI_IO_REBALANCE);
+                return Future.succeededFuture(clusterRebalance.getStatus());
+            case approve:
+                return requestRebalance(reconciliation, host, apiClient, false, rebalanceOptionsBuilder);
+            case refresh:
+                return requestRebalance(reconciliation, host, apiClient, true, rebalanceOptionsBuilder);
```
logging for these two cases?

tombentley commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaRebalanceAssemblyOperator.java:
```diff
@@ -0,0 +1,869 @@
+    }
+
+    /**
+     * Return the {@code RebalanceAnnotation} enum value for the raw String value of the strimzio.io/rebalance annotation
```
```suggestion
     * Return the {@code RebalanceAnnotation} enum value for the raw String value of the strimz.io/rebalance annotation
```

tombentley commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaRebalanceAssemblyOperator.java:
```diff
@@ -0,0 +1,869 @@
+     * Return the {@code RebalanceAnnotation} enum value for the raw String value of the strimzio.io/rebalance annotation
+     * set on the provided KafkaRebalance resource instance.
+     * If the annotation is not set it returns {@code RebalanceAnnotation.none} while if it's a not valid value, it
+     * returns {@code RebalanceAnnotation.unknown}
```
```suggestion
     * returns {@code RebalanceAnnotation.unknown}.
```

tombentley commented on cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/KafkaRebalanceAssemblyOperatorTest.java:
```diff
@@ -0,0 +1,167 @@
+import static org.mockito.Mockito.when;
+
+@ExtendWith(VertxExtension.class)
+public class KafkaRebalanceAssemblyOperatorTest {
```
Is this really the only test directly covering all that code in the `KafkaRebalanceAssemblyOperator`, or did I miss something?

tomncooper replied:
I am working on tests in a separate PR

ppatierno commented:
@scholzj regarding the metrics point, I verified that only the "real" reconcile is counted (i.e. `strimzi_reconciliations_total{kind="KafkaRebalance",} 2.0`) and not the internal reconcile.

scholzj commented:
@ppatierno Thanks. Lets see how many people will be confused by the failed to acquire lock errors. Is this suitable for another review pass?

ppatierno commented:
@scholzj yes it is. thanks.

ppatierno commented:
@strimzi-ci run tests

scholzj commented:
Any reason why not run it on Azure?

ppatierno commented:
/azp run system-tests

samuel-hawker commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/model/CruiseControlConfiguration.java:
```diff
@@ -40,23 +39,31 @@
-        CC_DEFAULT_PROPERTIES_MAP.put("completed.user.task.retention.time.ms", Long.toString(TimeUnit.DAYS.toMillis(1)));
-        CC_DEFAULT_PROPERTIES_MAP.put("default.goals", DEFAULT_GOALS);
-        CC_DEFAULT_PROPERTIES_MAP.put("goals", DEFAULT_GOALS);
+        CRUISE_CONTROL_DEFAULT_PROPERTIES_MAP = new HashMap<>();
```
Do we want to limit its size to 7?

tomncooper replied:
Converted it to an unmodifiable map.

samuel-hawker commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/model/CruiseControlConfiguration.java:
```diff
@@ -76,16 +83,7 @@ private CruiseControlConfiguration(String configuration, List<String> forbiddenO
-
     public static Map<String, String> getCruiseControlDefaultPropertiesMap() {
-        return CC_DEFAULT_PROPERTIES_MAP;
+        return CRUISE_CONTROL_DEFAULT_PROPERTIES_MAP;
```
Do we perhaps want to clone the map to stop any objects from modifying the defaults? (This might not be neccessary though...)

tomncooper replied:
I converted it to an unmodifiable map in the definition at the top of the file (using `CRUISE_CONTROL_DEFAULT_PROPERTIES_MAP = Collections.unmodifiableMap(config);`) this matches what we do with the zookeeper default config.

ppatierno replied:
@tomncooper remember to run checkstyle and mainly spotbugs. I had a problem with it. thanks.

samuel-hawker commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaRebalanceAssemblyOperator.java:
```diff
@@ -0,0 +1,904 @@
+ * <p>
+ *     This operator takes care of the {@code KafkaRebalance} custom resources that a user can create in order
+ *     to interact with Cruise Control REST API and execute a cluster rebalancing.
+ *     A state machine is used for the rebalacing flow which is reflected in the {@code status} of the custom resource.
```
rebalacing -> rebalancing

samuel-hawker commented on cluster-operator/src/test/java/io/strimzi/operator/cluster/model/CruiseControlTest.java:
```diff
@@ -723,6 +725,41 @@ public void testRestApiPortNetworkPolicy() {
+    @Test
+    public void testGoalsCheck() {
+
+        String customGoals = "com.linkedin.kafka.cruisecontrol.analyzer.goals.RackAwareGoal," +
```
I wonder, if we are re-using these strings, might we want to set each as a constant in the CruiseControl class to avoid string repetition?

tomncooper replied:
Did as you suggested and pointed all goal refs at these constants. Also made the default goal configs actual lists so we can programatically use them if we need to at a later date.

samuel-hawker commented on cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/resource/cruisecontrol/CruiseControlClientTest.java:
```diff
@@ -0,0 +1,166 @@
+
+        client.rebalance(HOST, PORT, rbOptions, MockCruiseControl.REBALANCE_NOT_ENOUGH_VALID_WINDOWS_ERROR)
+                .setHandler(result -> {
+                    if (result.succeeded()) {
```
context.succeeding and use checkpoint flags

samuel-hawker commented on cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/resource/cruisecontrol/CruiseControlClientTest.java:
```diff
@@ -0,0 +1,166 @@
+        Checkpoint checkpoint = context.checkpoint(6);
+        client.rebalance(HOST, PORT, rbOptions, null).setHandler(context.succeeding(result -> {
+            context.verify(() -> assertThat(result.getUserTaskId(), is(MockCruiseControl.REBALANCE_NO_GOALS_VERBOSE_RESPONSE_UTID)));
+            checkpoint.flag();
```
This many flags is unnecessary I believe one at the end is sufficient

samuel-hawker commented on cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/resource/cruisecontrol/CruiseControlClientTest.java:
```diff
@@ -0,0 +1,166 @@
+            context.verify(() -> assertThat(result.getUserTaskId(), is(MockCruiseControl.USER_TASK_REBALANCE_NO_GOALS_VERBOSE_RESPONSE_UTID)));
+            checkpoint.flag();
+            context.verify(() -> assertThat(result.getJson().getJsonObject(CC_REST_API_SUMMARY), is(notNullValue())));
+            checkpoint.flag();
```
ditto

ppatierno commented:
Other than the Sam comments, I don't see more outstanding comments that need changes right now. Regarding the wrong references, we'll do a squash changing the commit message before merging of course.

scholzj commented:
> Other than the Sam comments, I don't see more outstanding comments that need changes right now.

Well, For example this one is still open AFAIK: https://github.com/strimzi/strimzi-kafka-operator/pull/2945/files#r421365789

> Regarding the wrong references, we'll do a squash changing the commit message before merging of course.

I don't think squashing solves that. You need to change the messages specifically. I think it would be good to see that done before the merging to make sure the references are gone.

ppatierno commented:
as I said "we'll do a squash changing the commit message" ... so changing the message :-)

ppatierno commented:
/azp run system-tests

ppatierno commented:
/azp run system-tests

ppatierno commented:
@strimzi-ci run tests

ppatierno commented:
Travis and Jenkins are green. Azure seems to have some problems. Agree with @Frawless we can merge it!
